cli/src/main/java/hudson/cli/PlainCLIProtocol.java

```diff
@@ -0,0 +1,354 @@
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.CountingInputStream;
+
+/**
+ * CLI protocol working over a plain socket-like connection, without SSH or Remoting.
+ * Each side consists of frames starting with an {@code int} length,
+ * then a {@code byte} opcode, then any opcode-specific data.
+ * The length does not count the length field itself nor the opcode, so it is nonnegative.
+ */
+class PlainCLIProtocol {
+
+    static final Logger LOGGER = Logger.getLogger(PlainCLIProtocol.class.getName());
+
+    /** One-byte operation to send to the other side. */
+    private enum Op {
+        /** UTF-8 command name or argument. */
+        ARG(true),
+        /** UTF-8 locale identifier. */
+        LOCALE(true),
+        /** UTF-8 client encoding. */
+        ENCODING(true),
+        /** Start running command. */
+        START(true),
+        /** Exit code, as int. */
+        EXIT(false),
+        /** Chunk of stdin, as int length followed by bytes. */
+        STDIN(true),
+        /** EOF on stdin. */
+        END_STDIN(true),
+        /** Chunk of stdout. */
+        STDOUT(false),
+        /** Chunk of stderr. */
+        STDERR(false);
+        /** True if sent from the client to the server; false if sent from the server to the client. */
+        final boolean clientSide;
+        Op(boolean clientSide) {
+            this.clientSide = clientSide;
+        }
+    }
+
+    static abstract class EitherSide implements Closeable {
+
+        private final CountingInputStream cis;
+        private final FlightRecorderInputStream flightRecorder;
+        final DataInputStream dis;
+        final DataOutputStream dos;
+
+        protected EitherSide(InputStream is, OutputStream os) {
+            cis = new CountingInputStream(is);
+            flightRecorder = new FlightRecorderInputStream(cis);
+            dis = new DataInputStream(flightRecorder);
+            dos = new DataOutputStream(os);
+        }
+
+        final void begin() {
+            new Reader().start();
+        }
+
+        private class Reader extends Thread {
+
+            Reader() {
+                super("PlainCLIProtocol"); // TODO set distinctive Thread.name
+            }
+
+            @Override
+            public void run() {
+                try {
+                    while (true) {
+                        LOGGER.finest("reading frame");
+                        int framelen;
+                        try {
+                            framelen = dis.readInt();
+                        } catch (EOFException x) {
+                            handleClose();
+                            break; // TODO verify that we hit EOF immediately, not partway into framelen
+                        } catch (IOException x) {
+                            if (x.getCause() instanceof TimeoutException) {
```

Comment:
On Tomcat 8 it seems this is a `SocketTimeoutException`. Have not yet checked whether catching that (and continuing) allows a long-running `groovysh` or the like to work.

We could also send empty ping packets in either direction after every 15s or so of inactivity on that side, but there is no way to know what timeouts will be applied by different containers.

Reply:
[JENKINS-46659](https://issues.jenkins-ci.org/browse/JENKINS-46659)